### PR TITLE
Add ranges for warnings

### PR DESCRIPTION
**Pull request by adalinesimonian**
Partially addresses #5694

Adds range support for warnings.

**Work in progress, not ready for merge.**

Tasks:

- [x] Update PostCSS to 8.4
- [x] Update postcss-value-parser to 4.2
- [x] Update `Warning` type
- [x] Update `Problem` type
- [x] Update `report` function
- [x] Update some rules
- [x] Update tests for these rules

### CODE REVIEW

**Comment by adalinesimonian** [comment]
What this generally looks like in VS Code with the appropriate changes downstream:

Screenshot of VS Code reporting Stylelint warnings with ranges

**Comment by jeddy3** [comment]
You're doing amazing work extending the capabilities of both PostCSS and Stylelint with this range stuff!

**Comment by jeddy3** [comment]
(Remove the label to be consistent with [our](https://stylelint.io/maintainer-guide/pull-requests) [usage](https://stylelint.io/maintainer-guide/issues#labels), i.e. only `pr: *` labels for pull requests.)

**Comment by adalinesimonian** [comment]
I was running into issues trying to determine the end position for functions and other nodes provided by postcss-value-parser. I've opened up a PR there that will resolve this problem: TrySound/postcss-value-parser#83

**Comment by adalinesimonian** [comment]
Just an update: I think I'm somewhere halfway through all of the rules. It's mind-numbing tedium so I'm slowing down a bit since I started on this. But I'm still seeing it through.

The issue I'm thinking of is that it is going to be hard to review this PR since it will change almost every rule's implementation. While most rules may only have a couple of changed lines, some have very thoroughly reworked internals.

My setup for working on this PR has been to have the VS Code extension running in a workspace where I enable different rules as I work on them and test what ranges are reported to the editor. This makes it way easier to test. I was thinking about maybe creating a repo that has a folder for each rule and its different supported configuration options. This way someone could pull it down, and using a modified version of the extension, test the reported ranges for each rule.

Would this be helpful? My only hesitation is that it would take a bit of tedious work to create this thing. But then again, this whole PR is a bit of a pain in the butt so what's a little bit more if it saves people some time and hassle?

**Comment by ybiquitous** [comment]
@adalinesimonian Thank you so much for this big change! Your suggestion seems interesting, but I'm not sure how the maintenance cost of such repositories will be.

**Comment by adalinesimonian** [comment]
> @adalinesimonian Thank you so much for this big change! Your suggestion seems interesting, but I'm not sure how the maintenance cost of such repositories will be.

It'd be temporary, not a long-term thing. If someone wanted to keep it updated, they could, but I'm mostly focused on trying to get this PR reviewed thoroughly.

**Comment by adalinesimonian** [comment]
[stylelint-rules-test](https://github.com/adalinesimonian/stylelint-rules-test) is now up, from which you can inspect how errors are reported for rules easily.

Currently, only a few rules are up, more are being added. Instructions on how to use the repository are provided in the [readme](https://github.com/adalinesimonian/stylelint-rules-test#readme).

**Comment by adalinesimonian** [comment]
I'm feeling a little unwell, so I'll be away from this PR for a little bit. I hope to get back to it shortly!

**Comment by ybiquitous** [comment]
@adalinesimonian Please take care of yourself!

Just a question. Can we extract only the update of PostCSS into another PR?

**Comment by adalinesimonian** [comment]
> @adalinesimonian Please take care of yourself!
> 
> Just a question. Can we extract only the update of PostCSS into another PR?

Feel free!

**Comment by jeddy3** [comment]
@adalinesimonian I hope you feel better soon.

I'm just looping back to this pull request now - I've had little time for OSS these past few weeks, but I'll have time in the new year (after going offline for 10 days of Xmas). I'll be able to help more with this pull request then (updating rules, testing, reviewing and whatnot).

Thanks again for working on this range stuff, by the way! It vastly improves the experience for the user in the editor.

Am I right in thinking that this change is non-breaking and that each rule can opt-in to the range approach? If that's the case, could we break up the work of migrating the rules over, i.e. we create an umbrella task-list issue that we can pick away at? We could then incrementally improve the quality of our reporting over time.

File: package.json
Comment (ybiquitous) [inline package.json:136]: I've updated `postcss` via PR #5778. Could you rebase the main branch?

**Comment by ybiquitous** [comment]
I agree with @jeddy3 suggestion. I'd like to make progress on this wonderful feature if I can help. 💪🏼

**Comment by jeddy3** [comment]
> I'd like to make progress on this wonderful feature if I can help

Fantastic. 

I suggest we:

1. Get this pull request ready to merge
2. Systematically move through the remaining rules

For the latter, I've converted https://github.com/stylelint/stylelint/issues/5694 into an umbrella issue and added a task list. I suggest we don't update any more stylistic rules.

For the former, I suggest:

1. We resolve the conflicts
2. Update the tests for the rules already adapted in this pull request.

We need to be pragmatic about this. It's not practical to add `endLine` and `endColumn` properties to every reject test as we've thousands. I suggest adding them to just a few choice tests. I also suggest we revert any changes to the stylistic rules in the pull request if they require a lot more work to finish.

We can always patch mistakes if we get the positions wrong for a rule. We don't have the time or resources to get it perfect first time around. 

What do you think about this approach?

**Comment by ybiquitous** [comment]
@jeddy3 Your suggestion looks so nice. 👍🏼

**Comment by ybiquitous** [comment]
I'm going to fix the failed tests. 👀

File: lib/rules/declaration-no-important/index.js
Comment (ybiquitous) [inline lib/rules/declaration-no-important/index.js:36]: [note] It may be better to revert the change for `declaration-no-important` and fix it later (via another pull request).

**Comment by ybiquitous** [comment]
@jeddy3 Does it make sense to mark stylistic rules *deprecated* (adding `meta.deprecated = true`) before advancing this pull request? I want to revert the changes in stylistic rules to reduce the diff.

**Comment by ybiquitous** [comment]
I see. Thanks for the answer. 👍🏼

**Comment by ybiquitous** [comment]
Reverted frozen stylistic rules via faeb583.

File: lib/rules/declaration-no-important/index.js
Comment (ybiquitous) [reply lib/rules/declaration-no-important/index.js]: I've fixed the `!important` position via 85540e6.

File: lib/utils/getImportantPosition.js
Comment (ybiquitous) [inline lib/utils/getImportantPosition.js:15]: [note] I cannot find how to get a position of `!important` from CSS source code via `postcss` or `postcss-value-parser`. So, I've written the tiny utility `getImportantPosition`.

If anyone knows a better way, please let me know.

File: lib/rules/function-calc-no-unspaced-operator/index.js
Comment (ybiquitous) [inline lib/rules/function-calc-no-unspaced-operator/index.js:76]: [ask] Honestly, I doubt if this change is valuable. The previous report seems natural to me. Any thoughts?

(Screenshots on VS Code ⬇️ )

### Before

image

### After

image

Comment (jeddy3) [reply lib/rules/function-calc-no-unspaced-operator/index.js:76]: Before seems more natural. Let's revert.

Comment (ybiquitous) [reply lib/rules/function-calc-no-unspaced-operator/index.js:76]: Thanks. I'll do it. 👍🏼

File: lib/rules/declaration-block-single-line-max-declarations/index.js
Comment (ybiquitous) [inline lib/rules/declaration-block-single-line-max-declarations/index.js:51]: [ask] The previous version seems more useful to me because the rule is about a block `{ }` (especially when a selector is long). Any thoughts?

(Screenshots on VS Code ⬇️ )

### Before

image

### After

image

File: lib/rules/function-calc-no-unspaced-operator/index.js
Comment (ybiquitous) [reply lib/rules/function-calc-no-unspaced-operator/index.js]: Fixed via ec84751f

File: lib/rules/declaration-block-single-line-max-declarations/index.js
Comment (jeddy3) [reply lib/rules/declaration-block-single-line-max-declarations/index.js:51]: I agree. Let's revert.

The `endIndex` should point to the closing brace.

> because the rule is about a block { } (especially when a selector is long). 

Yes, that's correct. The majority of the rules [are named](https://stylelint.io/user-guide/rules/about#names) so that the first part describes the scope of the rule, e.g. [`declaration-property-unit-allowed-list`](https://stylelint.io/user-guide/rules/list/declaration-property-unit-allowed-list) is named as such (where the scope is `declaration`) rather than named `property-unit-allowed-list`. I think we can generally follow this pattern when deciding where the `index` and `endIndex`s should be.

Comment (ybiquitous) [reply lib/rules/declaration-block-single-line-max-declarations/index.js:51]: @jeddy3 Thank you. I understand well. 👍🏼

Comment (ybiquitous) [reply lib/rules/declaration-block-single-line-max-declarations/index.js:51]: Fixed via 19253dd0

**Comment by ybiquitous** [comment]
I've removed the unused utility functions below:

- `getWhitespaceStartIndex()` 2c596890
- `getWhitespaceEndIndex()` 9c675ce3
- `getWordStartIndex()` acca7dd9
- `getWordEndIndex()` 2755c1b8

Let's add them again if they are necessary.

**Comment by ybiquitous** [comment]
All the rules have not been supported yet, but this pull request is ready to review. 🎉 
(we need to address the remaining rules on other pull requests.)

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@ybiquitous Thank you so much for picking this up! You've done fantastic work on top of @adalinesimonian's. 

All the code LGTM. Let's update our docs too. I think we have a couple of examples that need updating [here](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/formatters.md), [here](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/plugins.md#testing) and [here](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/plugins.md#stylelintutilscheckagainstrule).

I think once that's done we can merge and release.

> All the rules have not been supported yet, but this pull request is ready to review.

I agree. When it comes to the other rules, I don't think we need to add end positions for all the reject tests. We should add just enough of them to be confident that the extra `endIndex` logic is working.

**Comment by ybiquitous** [comment]
Ah, thanks for suggesting updating the docs. I'll do it soon. 👍🏼

**Comment by ybiquitous** [comment]
Updating the docs was done via 13030a9.

**Review by jeddy3 — APPROVED** [review]
Fantastic work, thank you!

LGTM.

Feel free to merge. Any idea for the changelog entry? Perhaps:

- Added: ranges for warnings that can be used by formatters and integrations

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for the review and suggestion! I'll do it.

**Comment by ybiquitous** [comment]
Done.

> - Added: ranges for warnings that can be used by formatters and integrations ([#5725](https://github.com/stylelint/stylelint/pull/5725)).

**Pull request merged**
